File: C++/문제_13549번.cpp
#include <iostream>
#include <algorithm>
using namespace std;

//문제 13549번
#include <queue>
queue<int>q;
int dp[100001];
bool visit[100001];
int dx[3] = {2, 1 ,-1}, from , to;
void bfs(int from);
int main(){
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);
    cout.tie(NULL);
    cin >> from >> to;
    fill_n(dp, 100001, 987654321);
    q.push(from);
    bfs(from);
    cout << dp[to];
}
//bfs로 풀면 가중치를 못 쳐주어 dx의 인덱스를 2 1 -1로 해둬야 한다.
void bfs(int from){
    dp[from] = 0;
    while(!q.empty()){
        int pos = q.front();
        q.pop();
        for(int i = 0;i<3;i++){
            int go;
            int val;
            if(i != 0){
                go = pos + dx[i];
                val = dp[pos] + 1;
            }
            else{
                go = pos * dx[i];
                val = dp[pos];
            }
            if(go >= 0 && go <= 100000 && val < dp[go]){
                dp[go] = val;
                q.push(go);
            }
        }
    }
}

File: JavaScript/문제_13549번.js
let input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");
class Queue {
  constructor() {
    this.storage = {};
    this.front = 0;
    this.rear = 0;
  }

  size() {
    if (this.storage[this.rear] === undefined) {
      return 0;
    } else {
      return this.rear - this.rear + 1;
    }
  }

  push(value) {
    if (this.size() === 0) {
      this.storage['0'] = value;
    } else {
      this.rear += 1;
      this.storage[this.rear] = value;
    }
  }

  pop() {
    let temp;
    if (this.front === this.rear) {
      temp = this.storage[this.front];
      delete this.storage[this.front];
      this.front = 0;
      this.rear = 0;
    } else {
      temp = this.storage[this.front];
      delete this.storage[this.front];
      this.front += 1;
    }
    return temp;
  }

  front() {
    return this.storage[front];
  }
}
input = input[0].split(' ');
let start = +input[0];
let end = +input[1];
let visit = new Array(100001).fill(+999999999);
let dis = [1, -1, 2];
const bfs = (node) => {
  let q = new Queue();
  q.push(node);
  visit[node] = 0;
  while (q.size() > 0) {
    let now = q.pop();
    for (let i = 0; i < 3; i++) {
      let next = now + dis[i];
      if (i == 2) {
        next = now * 2;
        if (visit[next] > visit[now]) {
          q.push(next);
          visit[next] = visit[now];
        }
      }
      else if (next <= 100000 && next >= 0 && visit[next] > visit[now] + 1) {
        q.push(next);
        visit[next] = visit[now] + 1;
      }
    }
  }
}
bfs(start);
console.log(visit[end]);